### PR TITLE
Strong typed models

### PR DESCRIPTION
**Pull request by hardyscc**
#### Schema
```ts
export interface User extends Document {
	id: string;
	name: string;
	age: number;
}

export const UserTypedModel = dynamoose.model<User>(
	"User",
	new dynamoose.Schema({
		"id": String,
		"name": {
			"type": String,
			"index": {
				"global": true
			}
		},
		"age": Number
	})
);
```

#### Model
```ts
// Code here
```

#### General
```ts
const shouldCreateSuccessfully = UserTypedModel.create({"id": "1", "name": "john", "age": 25});

const shouldNotFailWithNotAllAttributesPassedIn = UserTypedModel.create({"id": "1"});

//@ts-expect-error
const shouldFailWithInvalidAttributes = UserTypedModel.create({"id": "1", "random": "string"});

const shouldQuerySuccessfully = UserTypedModel.query("name").eq("john").exec();

const shouldScanSuccessfully = UserTypedModel.scan("name").eq("john").exec();
```

### GitHub linked issue:

#934 

### Type (select 1):
- [ ] Bug fix
- [x] Feature implementation
- [ ] Documentation improvement
- [x] Testing improvement

- [ ] Test added to report bug (GitHub issue #--- @---)
- [ ] Something not listed here

### Is this a breaking change? (select 1):
- [ ] 🚨 YES 🚨
- [x] No
- [ ] I'm not sure

### Is this ready to be merged into Dynamoose? (select 1):
- [x] Yes
- [ ] No

### Are all the tests currently passing on this PR? (select 1):
- [x] Yes
- [ ] No

### Other:
- [x] I have read through and followed the Contributing Guidelines
- [x] I have searched through the GitHub pull requests to ensure this PR has not already been submitted
- [x] I have updated the Dynamoose documentation (if required) given the changes I made
- [x] I have added/updated the Dynamoose test cases (if required) given the changes I made
- [x] I have ensured the following commands are successful from the root of the project directory
  - [x] `npm test`
  - [x] `npm run lint`
- [x] I agree that all changes made in this pull request may be distributed and are made available in accordance with the [Dynamoose license](https://github.com/dynamoose/dynamoose/blob/master/LICENSE)
- [x] All of my commits and commit messages are detailed, explain what changes were made, and are easy to follow and understand
- [x] I have filled out all fields above

### CODE REVIEW

File: test/types/model/Query.ts [hunk 1]
```diff
@@ -0,0 +1,5 @@
+/* eslint @typescript-eslint/no-unused-vars: 0 */
+
+import {UserTypedModel} from "../Model";
+
+const shouldQuerySuccessfully = UserTypedModel.query("name").eq("john").exec();
```
Comment (fishcharlie) [inline test/types/model/Query.ts:5]: Shouldn't we add tests for the `lt`, `le`, etc. methods here?

File: test/types/model/Scan.ts [hunk 2]
```diff
@@ -0,0 +1,5 @@
+/* eslint @typescript-eslint/no-unused-vars: 0 */
+
+import {UserTypedModel} from "../Model";
+
+const shouldScanSuccessfully = UserTypedModel.scan("name").eq("john").exec();
```
Comment (fishcharlie) [inline test/types/model/Scan.ts:5]: Shouldn't we add tests for the `lt`, `le`, etc. methods here?

File: lib/DocumentRetriever.ts [hunk 3]
```diff
@@ -19,7 +19,7 @@ interface DocumentRetrieverTypeInformation {
 
 // DocumentRetriever is used for both Scan and Query since a lot of the code is shared between the two
 // type DocumentRetriever = BasicOperators;
-interface DocumentRetriever extends BasicOperators {} // eslint-disable-line @typescript-eslint/no-empty-interface
+// interface DocumentRetriever extends BasicOperators {} // eslint-disable-line @typescript-eslint/no-empty-interface
```
Comment (fishcharlie) [inline lib/DocumentRetriever.ts:22]: I think this can be removed now.

File: test/types/Model.ts [hunk 4]
```diff
@@ -9,3 +10,24 @@ const shouldFailWithNumberAsName = dynamoose.model(1);
...
 
 const shouldSucceedWithOnlyPassingInName = dynamoose.model("User");
+
+// Typed Models
+export interface User extends Document {
+	id: string;
+	name: string;
+	age: number;
+}
+
```
Comment (fishcharlie) [inline test/types/Model.ts:20]:
```suggestion
```

**Comment by hardyscc** [comment]
Thanks @fishcharlie,  Im a bit busy right now, will work on it few days later ^^

File: test/types/model/Query.ts (hunk 1 above)
Comment (hardyscc) [reply test/types/model/Query.ts]: added more tests

File: test/types/model/Scan.ts (hunk 2 above)
Comment (hardyscc) [reply test/types/model/Scan.ts]: added more tests

File: lib/DocumentRetriever.ts (hunk 3 above)
Comment (hardyscc) [reply lib/DocumentRetriever.ts]: deleted

File: test/types/Model.ts (hunk 4 above)
Comment (hardyscc) [reply test/types/Model.ts:20]: fixed

**Comment by hardyscc** [comment]
Hi @fishcharlie just updated the source code accordingly, please help review. 
We definitely need some kind of documentation, let me think about how to document it out. ^^

**Comment by geongun20** [comment]
This PR should be merged as quickly as possible...

**Comment by fishcharlie** [comment]
@geongun20 Agreed. However, Dynamoose is a volunteer project for me. This isn't my full time job. I'd love to have unlimited resources and time to work on this project, however, that isn't the reality of it.

You are more than welcome to review this PR, use this branch, help test it, donate to the project, etc. But besides that, please recognize that I have limited time available to work on this project. Maybe someday donations to the project will be large enough to change that, but that isn't the reality now.

For more information you can see this post I made in the Dynamoose Slack channel recently: https://dynamoose.slack.com/archives/CG2TSBML2/p1597687051311100.

I **hope** to get around to reviewing this PR soon, as I want it to be merged in as well.

**Review by fishcharlie — CHANGES_REQUESTED** [review]
Still needs documentation

**Comment by hardyscc** [comment]
@fishcharlie I just added some documentation under Model.md, please see if this is appropriate.

File: docs/docs/guide/Model.md [hunk 5]
```diff
@@ -1,19 +1,34 @@
 The Model object represents a table in DynamoDB. It takes in both a name and a schema and has methods to retrieve, and save documents in the database.
 
-## dynamoose.model(name, [schema][, config])
+## dynamoose.model[&lt;Document&gt;](name, [schema][, config])
 
 This method is the basic entry point for creating a model in Dynamoose. When you call this method a new model is created, and it returns a Document initializer that you can use to create instances of the given model.
 
+An optional class which extends `Document` can be given right before the function bracket. This provide type checking when using operations like Model.create().
```
Comment (fishcharlie) [inline docs/docs/guide/Model.md:7]:
```suggestion
An optional TypeScript class which extends `Document` can be provided right before the function bracket. This provides type checking when using operations like `Model.create()`.
```

File: test/types/Model.ts (hunk 4 above)
Comment (fishcharlie) [reply test/types/Model.ts:20]:
```suggestion
```

**Review by fishcharlie — CHANGES_REQUESTED** [review]
Looks really good! Few more minor things and I'll merge.

**Comment by hardyscc** [comment]
> Looks really good! Few more minor things and I'll merge.

Thanks a lot ^^

**Review by fishcharlie — APPROVED** [review]

**Comment by hardyscc** [comment]
Just remind my self here, cannot close hardyscc/nestjs-dynamoose#76 
still need generic type support for Key.

**Comment by lmanerich** [comment]
I assume it would also need to export ScanResponse and QueryResponse so we are able to type the exec():

```typescript
const shouldScanSuccessfully: ScanResponse<UserTypedModel[]> = UserTypedModel.scan("name").eq("john").exec();
```

**Comment by TranBaVinhSon** [comment]
I even can't type `exex()` on scan and query. I'm not sure dynamoose is working well with Typescript

**Comment by hardyscc** [comment]
@TranBaVinhSon please try this PR by
```
npm install dynamoose/dynamoose#master
```

**Comment by hardyscc** [comment]
> I assume it would also need to export ScanResponse and QueryResponse so we are able to type the exec():
> 
> ```ts
> const shouldScanSuccessfully: ScanResponse<UserTypedModel[]> = UserTypedModel.scan("name").eq("john").exec();
> ```
> 
> Or any way to get the typed array?
> 
> ```ts
> const shouldScanSuccessfully: UserTypedModel[] = UserTypedModel.scan("name").eq("john").exec().resultList();
> ```

@lmanerich you just missed out `await`
```ts
const shouldScanSuccessfully = await UserTypedModel.scan("name").eq("john").exec();
```

**Comment by lmanerich** [comment]
> 
> @lmanerich you just missed out `await`
> 
> ```ts
> const shouldScanSuccessfully = await UserTypedModel.scan("name").eq("john").exec();
> ```

Doing in a single line like you sugestted it, result in:

```
Property 'count' does not exist on type 'Condition'.ts(2339)
```

Doing in multiples lines like:

```
const query = await UserTypedModel.query('name');
query.eq("john");
const result = await query.exec();
console.log(result.count);
const user: UserTypedModel = result[0]; // results in Type 'Document[]' is missing the following properties from type 'UserTypedModel': name, email, lastName, and 10 more.
```

**Comment by hardyscc** [comment]
@lmanerich Yes

@fishcharlie I think we should export this two interfaces under `DocumentRetriever.ts` as well
```
interface ScanResponse<T> extends DocumentRetrieverResponse<T> {
	scannedCount: number;
	timesScanned: number;
}
interface QueryResponse<T> extends DocumentRetrieverResponse<T> {
	queriedCount: number;
	timesQueried: number;
}
```

**Comment by TranBaVinhSon** [comment]
@hardyscc 

I don't know why but `exec()` of `#master` version always return `{count: 0}` even though the item exist on dynamodb

**Comment by hardyscc** [comment]
@troyswanson I just did a quick test on AWS, no problem at all, could you provide more details?
this PK only change the types, shouldn't have any runtime changes.

**Comment by fishcharlie** [comment]
@TranBaVinhSon Does this issue occur on v2.3.0? Or only the master branch? Could you create a MCVE and create an issue please?

**Comment by fishcharlie** [comment]
@hardyscc Quick question about this work. I have the following code:

```ts
class ItemObjectDB extends Document implements ItemObject {
	id: string | undefined;
	type: "point" | "area" | "entity";
}

const schema = new dynamoose.Schema({
	"id": {
		"type": String,
		"required": true,
		"default": uuidv4()
	},
	"type": {
		"type": String,
		"required": true,
		"enum": ["point", "area", "entity"]
	}
});
```

But TypeScript complains that: `Property 'type' has no initializer and is not definitely assigned in the constructor`.

According to the documentation you have for this PR this should be valid. However, even when running the code you put in the documentation I'm getting TypeScript errors:

image

I think we need to update the documentation around this work. But I'm not sure what the solution is here. Do you have any thoughts or solutions here?

**Comment by hardyscc** [comment]
@fishcharlie I just try that out, seem no such typescript error.

Screenshot 2021-03-15 at 1 38 09 PM

**Comment by fishcharlie** [comment]
@hardyscc It looks like this only occurs when `strict` is set to `true` in the tsconfig.json file.

**Comment by hardyscc** [comment]
@fishcharlie oh yes correct!
In addition you can just set `strictPropertyInitialization` to `false` to remove that warning.

**Comment by lmanerich** [comment]
@fishcharlie @hardyscc Usually I declare the Documents as Interfaces, not classes. Doing that way, there is no error even with strict = true.

**Comment by fishcharlie** [comment]
@lmanerich Mind providing an example of that?

**Comment by lmanerich** [comment]
@fishcharlie 

```javascript
import { Document, dynamooseService, ModelType } from '@src/services/dynamoose.service';

export interface IAlert extends Document {
    id: string;
    email: string;
}

export const AlertSchema = dynamooseService.schema({
    id: {
        type: String,
        hashKey: true,
    },
    email: {
        required: true,
        type: String,
    },
});

const AlertModel: ModelType<IAlert> = dynamooseService.model<IAlert>('alert', AlertSchema);
export default AlertModel;
```

When whenever I need to use:

```javascript
const query = AlertModel.query();
query.limit(pageSize);
if (nextPage) {
    query.startAt(nextPage);
}
const result = await query.exec();
```

The only "issue", while using v2.3.0 is that the result object is a QueryResponse interface, so I need to do this (not sure about newer versions)

```javascript
// Type 'QueryResponse<Document[]>' is not assignable to type 'IAlert[]'.
const alerts: IAlert[] = []; //result
for (let i = 0; i < result.count; i++) {
    const alert: IAlert = (result[i] as unknown) as IAlert;
    alerts.push(alert);
}
```

**Comment by nicolasdeory** [comment]
> The only "issue", while using v2.3.0 is that the result object is a QueryResponse interface, so I need to do this (not sure about newer versions)
> 
> ```js
> // Type 'QueryResponse<Document[]>' is not assignable to type 'IAlert[]'.
> const alerts: IAlert[] = []; //result
> for (let i = 0; i < result.count; i++) {
>     const alert: IAlert = (result[i] as unknown) as IAlert;
>     alerts.push(alert);
> }
> ```

@lmanerich I believe you can do `await query.exec() as QueryResponse<IAlert>`. After that you can do `result[i] as IAlert`. The `QueryResponse<T>` interface extends the `T` type so it has its members as well as a few helper functions.

**Comment by lmanerich** [comment]
Hey @nicolasdeory , thanks for the suggestion.
I think this is not available to the release 2.3.0

```
Type 'QueryResponse' is not generic.
````

**Comment by fishcharlie** [comment]
This PR was merged in for v2.4.0

**Pull request merged**